seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,227 @@
+    """
+    session = anvil_session_store.get_session(user)
+    if fields is None:
+        r = session.get("api/workspaces")
+    else:
+        r = session.get("api/workspaces", params = {"fields": fields})
+    if r.status_code != 200:
+        raise TerraAPIException(
+            'Error: called Terra API "api/workspaces" got status: {} with a reason: {}'.format(r.status_code, r.reason))
+    return json.loads(r.text)
+
+
+def get_anvil_workspace_acl(workspace_namespace, workspace_name):
+    """
+    Request FireCloud access control list for workspace.
+
+    Args:
+        workspace_namespace (str): namespace (name of billing project) of the workspace
+        workspace_name (str): the name of the workspace
+    Returns:
+        {
+            "user1Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user2Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user3Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            }
+          }
+          :param workspace_name:
+          :param workspace_namespace:
+    """
+    if workspace_namespace and workspace_name:
+        uri = "api/workspaces/{0}/{1}/acl".format(workspace_namespace, workspace_name)
+        r = anvil_session_store.service_account_session.get(uri)
+        if r.status_code in DEFAULT_REFRESH_STATUS_CODES:  # has failed in refreshing the access code
+            anvil_session_store.service_account_session = AnvilSession(
+                service_account_info = GOOGLE_SERVICE_ACCOUNT_INFO, scopes = SOCIAL_AUTH_GOOGLE_OAUTH2_SCOPE)
+            r = anvil_session_store.service_account_session.get(uri)  # retry with the new access code
+        if r.status_code != 200:
+            raise TerraAPIException(
+                'Error: called Terra API "{}" got status: {} with a reason: {}'.format(uri, r.status_code, r.reason))
+        return json.loads(r.text)['acl']
+    else:
+        return {}
+
+
+class AnvilSessionStore(object):
+    sessions = {}
+
+    def __init__(self):
+        self._service_account_session = AnvilSession(service_account_info = GOOGLE_SERVICE_ACCOUNT_INFO,
+                                                         scopes = SOCIAL_AUTH_GOOGLE_OAUTH2_SCOPE)
+
+    @property
+    def service_account_session(self):
+        return self._service_account_session
+
+    @service_account_session.setter
+    def service_account_session(self, session):
+        self._service_account_session = session
+
+    def get_session(self, user):
+        social = user.social_auth.filter(provider = 'google-oauth2') if user is not None else None
+        if not social:
+            return
```

Comment:
Okay. I was using `get_session()` to detect if the user logged-in with Google. It is not an exception if a non-Google signed-in user gets a None return. I'm adding a `has_anvil_session(user)` to replace the use of `get_session()` for detection. After that, I will update this function.

Reply:
then I would you make a named helper method to make that clearer, maybe something like `is_user_google_authenticated`. Then in `get_session` you could call that

Reply:
I have defined a helper function `is_user_google_authenticated()` which does a similar thing as `user.social_auth.filter(provider = 'google-oauth2')` and returns a bool to indicate if a social auth exists for the user. The function seems not suitable here because the bool output is not enough and I still need the result of `social`. I will ensure all the calling of `get_session()` must be Google-authenticated.
Or, I change to use `requests` and `Authentication Bearer` header then there is no session management anymore.